File: level2/gbmv.py
from __future__ import annotations

from exo import *
from exo.libs.memories import DRAM_STATIC
from exo.platforms.x86 import *
from exo.syntax import *
from exo.stdlib.scheduling import *

"""
GEMV with a general band matrix
"""

@proc
def gbmv_no_trans(
  alpha: f32,
  beta: f32,
  n: size,
  m: size,
  A: f32[n, m],
  x: f32[m],
  y: f32[n],
  kl : size,
  ku : size
):
  for i in seq(0, n):
      y[i] = beta * y[i]
      for j in seq(0, m):
          if i-kl <= j and j <= i+ku:
              if 0 <= j + kl - i and j + kl - i < m:
                  y[i] += alpha * A[i, j + kl - i] * x[j]

@proc
def gbmv_trans(
  alpha: f32,
  beta: f32,
  n: size,
  m: size,
  A: f32[n, m],
  x: f32[n],
  y: f32[m],
  kl : size,
  ku : size
):
  for i in seq(0, m):
      y[i] = beta * y[i]
      for j in seq(0, n):
          if i-ku <= j and j <= i+kl:
              if 0 <= j + ku - i and j + ku - i < n:
                  y[i] += alpha * A[j + ku - i, i] * x[j]


if __name__ == "__main__":
    print(gbmv_trans)
    print(gbmv_no_trans)

__all__ = ["gbmv_trans", "gbmv_no_trans"]

File: src/level2/test/gbmv.c
#include "gbmv.h"



#include <stdio.h>
#include <stdlib.h>



// gbmv_no_trans(
//     alpha : f32 @DRAM,
//     beta : f32 @DRAM,
//     n : size,
//     m : size,
//     a : f32[m, n] @DRAM,
//     x : f32[n] @DRAM,
//     y : f32[m] @DRAM,
//     kl : size,
//     ku : size
// )
void gbmv_no_trans( void *ctxt, const float* alpha, const float* beta, int_fast32_t n, int_fast32_t m, const float* a, const float* x, float* y, int_fast32_t kl, int_fast32_t ku ) {
for (int i = 0; i < m; i++) {
  y[(i) * (1)] = *beta * y[(i) * (1)];
  for (int j = 0; j < n; j++) {
    if (i - kl <= j && j <= i + ku) {
      if (0 <= j + kl - i && j + kl - i < n) {
        y[(i) * (1)] += *alpha * a[(i) * (n) + (j + kl - i) * (1)] * x[(j) * (1)];
      }
    }
  }
}
}

// gbmv_trans(
//     alpha : f32 @DRAM,
//     beta : f32 @DRAM,
//     n : size,
//     m : size,
//     a : f32[n, m] @DRAM,
//     x : f32[n] @DRAM,
//     y : f32[m] @DRAM,
//     kl : size,
//     ku : size
// )
void gbmv_trans( void *ctxt, const float* alpha, const float* beta, int_fast32_t n, int_fast32_t m, const float* a, const float* x, float* y, int_fast32_t kl, int_fast32_t ku ) {
for (int i = 0; i < m; i++) {
  y[(i) * (1)] = *beta * y[(i) * (1)];
  for (int j = 0; j < n; j++) {
    if (i - ku <= j && j <= i + kl) {
      if (0 <= j + kl - i && j + kl - i < n) {
        y[(i) * (1)] += *alpha * a[(j + kl - i) * (m) + (i) * (1)] * x[(j) * (1)];
      }
    }
  }
}
}

